storage/internal/benchmarks/main.go

```diff
@@ -494,6 +448,139 @@ func startRecordingResults(w io.Writer, g *errgroup.Group, oType outputType) {
 	})
 }
 
+type multiError []error
+
+func (e multiError) Error() string {
+	var sb strings.Builder
+
+	for _, err := range e {
+		sb.WriteString(err.Error())
+		sb.WriteRune('\n')
+	}
+
+	return sb.String()
+}
+
+// runSamples will run benchmarks until the required number of samples is
+// collected and recorded, there is a setup failure, or we hit the deadline
+func runSamples(ctx context.Context, opts *benchmarkOptions, out io.Writer) error {
+	multiErr := multiError{}
+	deadline, _ := ctx.Deadline()
+
+	// Record results async
+	recordResultGroup, _ := errgroup.WithContext(ctx)
+	startRecordingResults(out, recordResultGroup, opts.outType)
+
+	// Create a group to run benchmarks
+	benchGroup, ctx := errgroup.WithContext(ctx)
+
+	// Select concurrency
+	var concurrentBenchmarkRuns int
+	switch opts.workload {
+	default:
+		concurrentBenchmarkRuns = opts.numWorkers
+	case 6:
+		// Directory benchmarks parallelize on the object level, so only run one
+		// benchmark at a time
+		concurrentBenchmarkRuns = 1
+	}
+
+	benchGroup.SetLimit(concurrentBenchmarkRuns)
+
+	// Run benchmarks until the required number of samples is collected, or we
+	// hit the deadline
+	for i := 0; i < opts.numSamples && !time.Now().After(deadline); i++ {
+		benchGroup.Go(func() error {
+			var benchmark benchmark
+			switch opts.workload {
+			default:
+				benchmark = &w1r3{opts: opts, bucketName: opts.bucket}
+			case 6:
+				benchmark = &directoryBenchmark{opts: opts, bucketName: opts.bucket, numWorkers: opts.numWorkers}
+			}
+
+			if err := benchmark.setup(ctx); err != nil {
+				// If setup fails, it will probably continue failing.
+				// Returning the error here will cancel the context, which will
+				// stop the benchmarking.
+				return fmt.Errorf("run setup failed: %v", err)
+			}
+			if err := benchmark.run(ctx); err != nil {
+				// If a run fails, we continue, as it could be a temporary issue.
+				multiErr = append(multiErr, fmt.Errorf("run failed: %v", err))
+			}
+			if err := benchmark.cleanup(); err != nil {
+				// If cleanup fails, we continue, as a failure here is not critical.
+				// Cleanup may be expected to fail if there is an issue with the run.
+				multiErr = append(multiErr, fmt.Errorf("run cleanup failed: %v", err))
+			}
+			return nil
+		})
+	}
+
+	if err := benchGroup.Wait(); err != nil {
+		multiErr = append(multiErr, err)
+	}
+
+	close(results)
+	recordResultGroup.Wait()
+
+	if len(multiErr) > 0 {
+		return multiErr
+	}
+	return nil
+}
+
+// runInServerMode blocks forever servicing probe requests.
+// Number of samples requested is ignored in this mode.
+// Timeouts must be properly set at the probe level to ensure requests aren't
+// being serviced concurrently
+func runInServerMode(ctx context.Context, opts *benchmarkOptions) {
+	results = make(chan benchmarkResult, 4)
+
+	serverutils.Serve(func(request *epb.ProbeRequest, reply *epb.ProbeReply) {
+		timeLimit := time.Millisecond * time.Duration(request.GetTimeLimit())
```

Comment:
Maybe note that this is per-sample?

Reply:
Added a comment; let me know if it makes sense or is still missing context.